x-pack/plugins/rule_registry/server/rule_registry/rule_type_helpers/create_lifecycle_rule_type_factory.ts

```diff
@@ -0,0 +1,235 @@
+  TRuleParams extends RuleParams,
+  TActionVariable extends ActionVariable
+>(
+  type: RuleType<
+    TFieldMap,
+    TRuleParams,
+    TActionVariable,
+    { alertWithLifecycle: LifecycleAlertService<TFieldMap, TActionVariable> }
+  >
+) => RuleType<TFieldMap, TRuleParams, TActionVariable>;
+
+const trackedAlertStateRt = t.type({
+  alertId: t.string,
+  alertUuid: t.string,
+  started: t.string,
+});
+
+const wrappedStateRt = t.type({
+  wrapped: t.record(t.string, t.unknown),
+  trackedAlerts: t.record(t.string, trackedAlertStateRt),
+});
+
+export function createLifecycleRuleTypeFactory<
+  TRuleRegistry extends RuleRegistry<DefaultFieldMap>
+>(): TRuleRegistry extends RuleRegistry<infer TFieldMap>
+  ? CreateLifecycleRuleType<TFieldMap>
+  : never;
+
+export function createLifecycleRuleTypeFactory(): CreateLifecycleRuleType<DefaultFieldMap> {
+  return (type) => {
+    return {
+      ...type,
+      executor: async (options) => {
+        const {
+          services: { scopedRuleRegistryClient, alertInstanceFactory, logger },
+          state: previousState,
+          rule,
+        } = options;
+
+        const decodedState = wrappedStateRt.decode(previousState);
+
+        const state = isLeft(decodedState)
+          ? {
+              wrapped: previousState,
+              trackedAlerts: {},
+            }
+          : decodedState.right;
+
+        const currentAlerts: Record<
+          string,
+          UserDefinedAlertFields<DefaultFieldMap> & { 'kibana.rac.alert.id': string }
+        > = {};
+
+        const timestamp = options.startedAt.toISOString();
+
+        const nextWrappedState = await type.executor({
+          ...options,
+          state: state.wrapped,
+          services: {
+            ...options.services,
+            alertWithLifecycle: ({ id, fields }) => {
+              currentAlerts[id] = {
+                ...fields,
+                'kibana.rac.alert.id': id,
+              };
+              return alertInstanceFactory(id);
+            },
+          },
+        });
+
+        const currentAlertIds = Object.keys(currentAlerts);
+        const trackedAlertIds = Object.keys(state.trackedAlerts);
+        const newAlertIds = currentAlertIds.filter((alertId) => !trackedAlertIds.includes(alertId));
+
+        const allAlertIds = [...new Set(currentAlertIds.concat(trackedAlertIds))];
+
+        const trackedAlertStatesOfRecovered = Object.values(state.trackedAlerts).filter(
+          (trackedAlertState) => !currentAlerts[trackedAlertState.alertId]
+        );
+
+        logger.debug(
+          `Tracking ${allAlertIds.length} alerts (${newAlertIds.length} new, ${trackedAlertStatesOfRecovered.length} recovered)`
+        );
+
+        const alertsDataMap: Record<string, UserDefinedAlertFields<DefaultFieldMap>> = {
+          ...currentAlerts,
+        };
+
+        if (scopedRuleRegistryClient && trackedAlertStatesOfRecovered.length) {
+          const { events } = await scopedRuleRegistryClient.search({
```

Comment:
I'm wondering if rather than searching for this data, we could cache it in either the rule state or rule instance state, both of which are persisted in the rule SO.  Problem with storing it there is bulking up those SOs, which could cause general perf issues.

Reply:
I've considered it, but indeed, the amount of data that would have to be stored in the rule state would be significantly more than what currently is needed. Esp with 100+ alerts being fired for an execution. Plus, it's caching, and we all know cache invalidation sucks 😄